contracts/HoprStake2.sol

```diff
@@ -0,0 +1,443 @@
+                if (redeemedFactorValue < factor) {
+                    redeemedFactor[from][index] = tokenId;
+                }
+                emit Redeemed(from, tokenId, redeemedFactorValue < factor);
+                break;
+            }
+        }
+        if (index == boostIndex) {
+            // new type being redeemed.
+            redeemedFactor[from][boostIndex] = tokenId;
+            redeemedFactorIndex[from] += 1;
+            emit Redeemed(from, tokenId, true);
+        }
+
+        return IERC721Receiver(address(this)).onERC721Received.selector;
+    }
+
+    /**
+     * @dev Manually sync account's reward states
+     * @notice public function of ``_sync``.
+     * @param account address Account whose stake rewards will be synced.
+     */
+    function sync(address account) external {
+        _sync(account); 
+    }
+
+    /**
+     * @dev Sync rewards and claim them
+     * @notice public function of ``_sync`` + ``_claim``
+     * @param account address Account whose stake rewards will be synced and claimed.
+     */
+    function claimRewards(address account) external {
+        _sync(account); 
+        _claim(account);
+    }
+
+    /**
+     * @dev Unlock staking
+     * @param account address Account that staked tokens.
+     */
+    function unlock(address account) external {
+        require(block.timestamp > PROGRAM_END, "HoprStake: Program is ongoing, cannot unlock stake.");
+        uint256 actualStake = accounts[account].actualLockedTokenAmount;
+        _sync(account); 
+        accounts[account].actualLockedTokenAmount = 0;
+        totalLocked -= actualStake;
+        _claim(account);
+        // unlock actual staked tokens
+        IERC20(LOCK_TOKEN).safeTransfer(account, actualStake);
+        // unlock redeemed NFTs
+        for (uint256 index = 0; index < redeemedNftIndex[account]; index++) {
+            NFT_CONTRACT.transferFrom(address(this), account, redeemedNft[account][index]);
+        }
+        emit Released(account, actualStake);
+    }
+
+    /**
+     * @dev Reclaim any ERC20 token being accidentally sent to the contract.
+     * @param tokenAddress address ERC20 token address.
+     */
+    function reclaimErc20Tokens(address tokenAddress) external onlyOwner nonReentrant {
+        uint256 difference;
+        if (tokenAddress == LOCK_TOKEN) {
+            difference = IERC20(LOCK_TOKEN).balanceOf(address(this)) - totalLocked;
+        } else {
+            difference = IERC20(tokenAddress).balanceOf(address(this));
+        }
+        IERC20(tokenAddress).safeTransfer(owner(), difference);
+    }
+
+    /**
+     * @dev Reclaim any ERC721 token being accidentally sent to the contract.
+     * @param tokenAddress address ERC721 token address.
+     */
+    function reclaimErc721Tokens(address tokenAddress, uint256 tokenId) external onlyOwner nonReentrant {
+        IHoprBoost(tokenAddress).transferFrom(address(this), owner(), tokenId);
```

Comment:
If `owner` would transfer redeemed NFTs out, this would break the `unlock` function - right? Cause the NFTs are still registered in `redeemedNft` and therefore the transfer in the loop would fail (and revert I guess?). I guess there's 2 options:
1. Properly unregister the NFT here
2. Don't allow `owner` to reclaim the actual boost NFT

I guess (2) is easier and hence the preferred option.